libusb/os/darwin_usb.c

```diff
@@ -167,6 +167,60 @@ static enum libusb_error darwin_to_libusb (IOReturn result) {
   }
 }
 
+uint32_t get_running_version(void) {
+  static int32_t running_version = 0;
+  if (running_version == 0) {
+    char os_version_string[64];
+    size_t os_version_string_len = sizeof(os_version_string) - 1;
+
+    /* newer versions of macOS provide a sysctl for the OS version */
+    int ret = sysctlbyname("kern.osproductversion", os_version_string, &os_version_string_len, NULL, 0);
```

Comment:
I can't find `kern.osproductversion` in any man page or header, which gives me pause.  Are we sure it's public API?

I'm not a big fan of falling back to using the file system either, since that would probably run afoul of App Sandbox, and require apps that don't otherwise use the file system to whitelist that path.

Seems preferable to use `kern.osrelease` which works since forever. It would require shifting from Darwin versions to macOS versions, but the wikipedia macOS articles have tables mapping the two.

Reply:
It has to do with an edge case for `kern.osrelease` that was kind of nasty. It was something like there was a new version of IOKit but the darwin version did not change. The systems where this occurs (10.4.7 and 10.5.4) don't really matter much so it is probably ok to treat them as 10.4.0 and 10.5.0.

Apple has not documented `kern.osproductversion` but it is well understood and widely used. It is set at boot time by launchd. See:

https://github.com/apple/darwin-xnu/commit/5bbb823c13f3ab1ab58878f96b35433a29882676#diff-1daf70413d68262120e719485aac47dd7da773d2333383c9fe5fd34dc93a5ffcR1694

Apple has serious documentation issues and there is no longer a way to communicate these back to them. radar is no longer accessible outside of Apple.

I have no problem with making it fall back to `kern.osrelease` instead of a file for the older releases and just remove the point release granularity. I might have to retest given this code was tested in VMs going back to 10.4.x.